solidity/interface-gen-tool.md

```diff
@@ -0,0 +1,82 @@
+# Purpose
+
+<!-- This section is also sometimes called "Motivations" or "Goals". -->
+
+<!-- It is fine to remove this section from the final document,
+but understanding the purpose of the doc when writing is very helpful. -->
+
+Design document for an interface auto generator script. This removes the need to manually create and modify interfaces after the creation/modification of a contract.
+
+<!-- Describe the specific problem that the document is seeking to address as well
+as information needed to understand the problem and design space.
+If more information is needed on the costs of the problem,
+this is a good place to that information. -->
+
+## Context
+
+Interfaces have recently been introduced throughout the OP Stack codebase, resulting in significant improvements to the development workflow. This addition has dramatically reduced compile time after the modification of most contracts from several minutes to a few seconds. Furthermore, the introduction of interfaces also makes it easier for external developers to work with the OP Stack codebase.
+
+## Problem Statement
+
+However, this improvement comes with a trade-off, the creation and modification of interfaces is currently a manual process. When changes are made to a contract's ABI, its corresponding interface must be manually updated. This approach lacks scalability and efficiency, especially when making large changes and as the codebase grows and evolves.
+
+## Solutions Considered
+
+<!-- List out a short summary of each possible solution that was considered.
+Comparing the effort of each solution -->
+
+While using a script that leverages `forge inspect` and `cast interface` to automatically generate interfaces for all relevant contract files could be a potential solution, it has several limitations:
```

Comment:
Other limitations:

- Types are re-generated instead of being pulled from libraries when appropriate
- Contract types are turned into `address` types